src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsOnly',
+      _body:
+        // rhs should not execute
+        'let t = (a == 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 2);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsAndRhs',
+      _body:
+        // rhs should execute
+        'let t = (a != 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('binary_mixed')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#arithmetic-expr')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#logical-expr')
+  .desc('Tests order of evaluation of both arithmetic and logical binary expressions.')
+  .paramsSimple([
+    {
+      name: 'ArithAndLogical',
+      _body: 'return mul(&a, 10) - bool_to_i32(mul(&a, 10) == 200 && mul(&a, 10) == 2000);',
+      _result: 20 - 1,
+    },
+    {
+      name: 'LogicalAndArith',
+      _body: 'return bool_to_i32(mul(&a, 10) == 20 && mul(&a, 10) == 200) - mul(&a, 10);',
+      _result: 1 - 2000,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('call')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#function-calls')
+  .desc('Tests order of evaluation of call expressions.')
+  .paramsSimple([
+    {
+      name: 'OneSE', //
+      _body: 'return sub3(mul(&a, 10), a, 3);',
+      _result: 20 - 20 - 3,
+    },
+    {
+      name: 'AllSE',
+      _body: 'return sub3(mul(&a, 10), mul(&a, 10), mul(&a, 10));',
+      _result: 20 - 200 - 2000,
+    },
+    {
+      name: 'MiddleNotSE',
+      _body: 'return sub3(mul(&a, 10), a, mul(&a, 10));',
+      _result: 20 - 20 - 200,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('index_accessor')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#array-access-expr')
+  .desc('Tests order of evaluation of index accessor expressions.')
+  .paramsSimple([
+    {
+      name: 'LeftSE', //
+      _body: 'return arr2D[mul(&a, 2)][a];',
+      _result: 4 * 10 + 4,
+    },
+    {
+      name: 'RightSE', //
+      _body: 'return arr2D[a][mul(&a, 2)];',
+      _result: 2 * 10 + 4,
+    },
+    {
+      name: 'BothSE',
+      _body: 'return arr2D[mul(&a, 2)][mul(&a, 2)];',
+      _result: 4 * 10 + 8,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('assignment')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#assignment')
+  .desc('Tests order of evaluation of assignment statements.')
+  .paramsSimple([
+    {
+      name: 'ToArray1D',
+      _body:
+        'arr1D_zero[mul(&a, 2)] = mul(&a, 2);' + //
+        'return arr1D_zero[8];',
+      _result: 4,
+    },
+    {
+      name: 'ToArray2D',
+      _body:
+        'arr2D_zero[mul(&a, 2)][mul(&a, 2)] = mul(&a, 2);' + //
```

Comment:
Note: You can use backticks ` to do multi-line strings.

Reply:
Yes, but I liked the alignment better this way.